File: 46圆圈中最后剩下的数/Solution.cpp
class Solution {
public:
    int LastRemaining_Solution(int n, int m)
    {
        if (n==0)
            return -1;
        if (n==1)
            return 0;
        return (LastRemaining_Solution(n-1,m)+m)%n;
    }
};
/* 约瑟夫环的递归解法，每一轮的胜者都是初始位置加指定数m再除以n的余数
然后这个胜者的位置又变成了下一次报数的初始位置 */

File: 46圆圈中最后剩下的数/Solution.java
public class Solution {
    public int LastRemaining_Solution(int n, int m) {
         if(n < 1 || m < 1){
            return -1;
        }
        int last = 0;
        for(int i = 2; i <= n; i++){
            last = (last + m) % i;
        }
        return last;
    }
}
/* 数学归纳法，我们可以得到只有一个人玩游戏的时候结果是0
两个人的时候是一个人的结果加上m在对人数取余，后面届时如此
由此就可以递推出最后的结果 */